File: agent_protocol/src/vonk_agent_protocol/package_upgrade.py
"""Exact source-bound rollback authority for the current package transaction."""

from typing import Annotated, Literal

from pydantic import Field, model_validator

from .wire_model import Digest, WireModel


class PackageRollbackSource(WireModel):
    package_sha256: Digest
    package_signature: Annotated[str, Field(pattern=r"^[0-9a-f]{128}$")]
    package_version: Annotated[
        str, Field(pattern=r"^[0-9A-Za-z][0-9A-Za-z.+~:-]{0,127}$")
    ]
    binary_sha256: Digest
    helper_sha256: Digest


class PackageRollbackAuthority(WireModel):
    source: PackageRollbackSource
    attempt_nonce: Digest
    activation_deadline: int = Field(strict=True, ge=1)


class PackageActivationReceipt(WireModel):
    schema_version: Literal[2]
    node_id: Annotated[str, Field(pattern=r"^spk_[0-9a-f]{32}$")]
    source_package_sha256: Digest
    source_version: Annotated[
        str, Field(pattern=r"^[0-9A-Za-z][0-9A-Za-z.+~:-]{0,127}$")
    ]
    source_binary_sha256: Digest
    candidate_package_sha256: Digest
    candidate_version: Annotated[
        str, Field(pattern=r"^[0-9A-Za-z][0-9A-Za-z.+~:-]{0,127}$")
    ]
    candidate_binary_sha256: Digest
    attempt_nonce: Digest
    phase: Literal[
        "armed",
        "activation_failed",
        "acknowledged",
        "rolling_back",
        "rolled_back",
        "rollback_failed",
    ]
    created_at: int = Field(strict=True, ge=1)
    updated_at: int = Field(strict=True, ge=1)
    outcome: Annotated[str, Field(pattern=r"^[a-z_]{1,128}$")]

    @model_validator(mode="after")
    def ordered_timestamps(self) -> "PackageActivationReceipt":
        if self.updated_at < self.created_at:
            raise ValueError("activation receipt precedes its transaction")
        return self

File: rust/crates/vonk-agent-protocol/src/package_upgrade.rs
//! Current, exact source-bound package rollback authority.
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct PackageRollbackSource {
    pub package_sha256: String,
    pub package_signature: String,
    pub package_version: String,
    pub binary_sha256: String,
    pub helper_sha256: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct PackageRollbackAuthority {
    pub source: PackageRollbackSource,
    pub attempt_nonce: String,
    pub activation_deadline: i64,
}

impl PackageRollbackAuthority {
    pub fn valid(&self) -> bool {
        let hex = |value: &str, length| {
            value.len() == length
                && value
                    .bytes()
                    .all(|c| c.is_ascii_digit() || (b'a'..=b'f').contains(&c))
        };
        hex(&self.source.package_sha256, 64)
            && hex(&self.source.package_signature, 128)
            && hex(&self.source.binary_sha256, 64)
            && hex(&self.source.helper_sha256, 64)
            && hex(&self.attempt_nonce, 64)
            && self.activation_deadline > 0
            && self
                .source
                .package_version
                .as_bytes()
                .first()
                .is_some_and(u8::is_ascii_alphanumeric)
            && self.source.package_version.len() <= 128
            && self
                .source
                .package_version
                .bytes()
                .all(|c| c.is_ascii_alphanumeric() || b".+~:-".contains(&c))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum PackageActivationPhase {
    Armed,
    ActivationFailed,
    Acknowledged,
    RollingBack,
    RolledBack,
    RollbackFailed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct PackageActivationReceipt {
    pub schema_version: u8,
    pub node_id: String,
    pub source_package_sha256: String,
    pub source_version: String,
    pub source_binary_sha256: String,
    pub candidate_package_sha256: String,
    pub candidate_version: String,
    pub candidate_binary_sha256: String,
    pub attempt_nonce: String,
    pub phase: PackageActivationPhase,
    pub created_at: i64,
    pub updated_at: i64,
    pub outcome: String,
}
